<!doctype html>
<html>
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <title>VCFtools</title>

    <link rel="stylesheet" href="stylesheets/styles.css">
    <link rel="stylesheet" href="stylesheets/github-light.css">
    <script src="javascripts/scale.fix.js"></script>
    <meta name="viewport" content="width=device-width, initial-scale=1, user-scalable=no">
    <!--[if lt IE 9]>
    <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
  </head>
  <body>
    <div class="wrapper">
      <header>
        <h1 class="header">VCFtools</h1>
        <p class="header">A set of tools written in Perl and C++ for working with VCF files.</p>

        <ul>
        	<li><a href="index.html">Home</a></li>
        	<li><a href="examples.html">Documentation</a></li>
        	<li class="download"><a class="buttons" href="https://github.com/auton1/vcftools/zipball/master">Download ZIP</a></li>
         <li class="download"><a class="buttons" href="https://github.com/auton1/vcftools/tarball/master">Download TAR</a></li>
         <li><a class="buttons github" href="https://github.com/auton1/vcftools">View On GitHub</a></li>
        </ul>
      </header>
      <section>
	<h2>The Perl modules and scripts</h2>
	<p>VCFtools contains a Perl API (<a href="#Vcf.pm">Vcf.pm</a>) and a number
	of Perl scripts that can be used to perform common tasks with VCF files such as
	file validation, file merging, intersecting, complements, etc. The Perl tools
	support all versions of the VCF specification (3.2, 3.3, 4.0, 4.1 and 4.2), 
	nevertheless, the users are encouraged to use the latest versions VCFv4.1 or VCFv4.2.
	The VCFtools in general have been used mainly with diploid data, but the Perl tools
	aim to support polyploid data as well.	Run any of the Perl scripts with the <strong>--help</strong> switch to obtain more help.	</p>
	
	<p>
	Many of the <strong>Perl scripts require that the VCF files are compressed by <span class="cmd">bgzip</span>
	and indexed by <span class="cmd">tabix</span></strong> (both tools are part of the htslib/tabix package,
	available for <a href="http://www.htslib.org/">download here</a>).
	The VCF files can be compressed and indexed using the following commands
	<p class="codebox">bgzip my_file.vcf<br>
	tabix -p vcf my_file.vcf.gz
	</p>

	<h2>The tools</h2>
	<ul>
		<!--<li> <a href="#fill-aa">fill-aa</a>-->
		<li> <a href="#fill-an-ac">fill-an-ac</a>
		<li> <a href="#fill-fs">fill-fs</a>
		<li> <a href="#fill-ref-md5">fill-ref-md5</a>
		<li> <a href="#fill-rsIDs">fill-rsIDs</a>
		<li> <a href="#vcf-annotate">vcf-annotate</a>
		<li> <a href="#vcf-compare">vcf-compare</a>
		<li> <a href="#vcf-concat">vcf-concat</a>
		<li> <a href="#vcf-consensus">vcf-consensus</a>
		<li> <a href="#vcf-contrast">vcf-contrast</a>
		<li> <a href="#vcf-convert">vcf-convert</a>
		<li> <a href="#vcf-filter">vcf-filter</a>
		<li> <a href="#vcf-fix-newlines">vcf-fix-newlines</a>
		<li> <a href="#vcf-fix-ploidy">vcf-fix-ploidy</a>
		<li> <a href="#vcf-indel-stats">vcf-indel-stats</a>
		<li> <a href="#vcf-isec">vcf-isec</a>
		<li> <a href="#vcf-merge">vcf-merge</a>
		<li> <a href="#vcf-phased-join">vcf-phased-join</a>
		<li> <a href="#vcf-query">vcf-query</a>
		<li> <a href="#vcf-shuffle-cols">vcf-shuffle-cols</a>
		<li> <a href="#vcf-sort">vcf-sort</a>
		<li> <a href="#vcf-stats">vcf-stats</a>
		<li> <a href="#vcf-subset">vcf-subset</a>
		<li> <a href="#vcf-to-tab">vcf-to-tab</a>
		<li> <a href="#vcf-tstv">vcf-tstv</a>
		<li> <a href="#vcf-validator">vcf-validator</a>
		<li> <a href="#Vcf.pm">Vcf.pm</a>
	</ul><br>

<script type="text/javascript" src="http://ajax.googleapis.com/ajax/libs/jquery/1.3.2/jquery.min.js"></script>
<script type="text/javascript">
jQuery(document).ready(function() {
        jQuery(".usageText").hide();
        jQuery(".usageToggle").click(function() { jQuery(this).next(".usageText").slideToggle(100); });
    });
</script>

<h3><a name="fill-an-ac" class="Q">fill-an-ac</a></h3>
<p> Fill or recalculate AN and AC INFO fields.
</p>
<p class="codebox">zcat file.vcf.gz | fill-an-ac | bgzip -c &gt; out.vcf.gz</p>
<div class="usageBox"><span class="usageToggle">(Read more)</span><pre class="usageText">Usage: fill-an-ac [OPTIONS] &lt; in.vcf &gt;out.vcf
Options:
   -h, -?, --help                      This help message.
</pre></div>

<h3><a name="fill-fs" class="Q">fill-fs</a></h3>
<p> Annotates the VCF file with flanking sequence (INFO/FS tag) masking known variants with N's. Useful for designing primers.
</p>
<p class="codebox">fill-fs -r /path/to/refseq.fa | vcf-query '%CHROM\t%POS\t%INFO/FS\n' &gt; out.tab</p>
<div class="usageBox"><span class="usageToggle">(Read more)</span><pre class="usageText">About: Annotate VCF with flanking sequence (INFO/FS tag)
Usage: fill-fs [OPTIONS] file.vcf
Options:
   -b, --bed-mask &lt;file&gt;           Regions to mask (tabix indexed), multiple files can be given
   -c, --cluster &lt;int&gt;             Do self-masking of clustered variants within this range.
   -l, --length &lt;int&gt;              Flanking sequence length [100]
   -m, --mask-char &lt;char|lc&gt;       The character to use or &quot;lc&quot; for lowercase. This option must preceed
                                       -b, -v or -c in order to take effect. With multiple files works
                                        as a switch on the command line, see the example below [N]
   -r, --refseq &lt;file&gt;             The reference sequence.
   -v, --vcf-mask &lt;file&gt;           Mask known variants in the flanking sequence, multiple files can be given (tabix indexed)
   -h, -?, --help                  This help message.
Example:
   # Mask variants from the VCF file with N's and use lowercase for the bed file regions
   fill-fs file.vcf -v mask.vcf -m lc -b mask.bed
</pre></div>

<h3><a name="fill-ref-md5" class="Q">fill-ref-md5</a></h3>
<p>Fill missing reference info and sequence MD5s into VCF header.
</p>
<p class="codebox">fill-ref-md5 -i "SP:Homo\ Sapiens" -r ref.fasta in.vcf.gz -d ref.dict out.vcf.gz</p>
<div class="usageBox"><span class="usageToggle">(Read more)</span><pre class="usageText">About: The script computes MD5 sum of the reference sequence and inserts
   'reference' and 'contig' tags into header as recommended by VCFv4.1.
   The VCF file must be compressed and tabix indexed, as it takes advantage
   of the lightning fast tabix reheader functionality.
Usage: fill-ref-md5 [OPTIONS] in.vcf.gz out.vcf.gz
Options:
   -d, --dictionary &lt;file&gt;             Where to read/write computed MD5s. Opened in append mode, existing records are not touched.
   -i, --info &lt;AS:xx,SP:xx,TX:xx&gt;      Optional info on reference assembly (AS), species (SP), taxonomy (TX)
   -r, --refseq &lt;file&gt;                 The reference sequence in fasta format indexed by samtools faidx
   -h, -?, --help                      This help message.
Examples:
   fill-ref-md5 -i AS:NCBIM37,SP:&quot;Mus\ Musculus&quot; -r NCBIM37_um.fa  -d NCBIM37_um.fa.dict in.vcf.gz out.vcf.gz
</pre></div>

<h3><a name="fill-rsIDs" class="Q">fill-rsIDs</a></h3>
<p>Fill missing rsIDs. This script has been discontinued, please use <a href="#vcf-annotate">vcf-annotate</a> instead.
</p>

<h3><a name="vcf-annotate" class="Q">vcf-annotate</a></h3>
<p> The script adds or removes filters and custom annotations to VCF files.
To add custom annotations to VCF files, create TAB delimited file with annotations
such as
<p class="codebox" style="white-space:pre;font-family:monospace;">#CHR     FROM   TO      ANNOTATION 
1        12345  22345   gene1 
1        67890  77890   gene2 
</p>
<p>
Compress the file (using <span class="cmd">bgzip annotations</span>), index (using <span class="cmd">tabix -s 1 -b 2 -e 3 annotations.gz</span>) and
run
</p>
<p class="codebox">cat in.vcf | vcf-annotate -a annotations.gz \ <br>
&nbsp;&nbsp; -d key=INFO,ID=ANN,Number=1,Type=Integer,Description='My custom annotation' \ <br>
&nbsp;&nbsp; -c CHROM,FROM,TO,INFO/ANN &gt; out.vcf </p>
<p style='padding-bottom:0px'>The script is also routinely used to apply filters. There are a number of predefined filters and 
custom filters can be easily added, see <span class="cmd">vcf-annotate -h</span> for examples. 
Some of the predefined filters take advantage of tags added by bcftools, the descriptions of the most frequently asked
ones follow:</p>
    <div style='margin-left:1em;margin-top:0.2em;'>
        <strong>Strand Bias</strong> 
        .. Tests if variant bases tend to come from one strand. Fisher's exact test
        for 2x2 contingency table where the row variable is being the reference allele
        or not and the column variable is strand. Two-tail P-value is used.
    </div>
    <div style='margin-left:1em;margin-top:0.2em;'>
        <strong>End Distance Bias</strong> 
        .. Tests if variant bases tend to occur
        at a fixed distance from the end of reads, which is usually an indication of
        misalignment. (T-test)
    </div>
    <div style='margin-left:1em;margin-top:0.2em;'>
        <strong>Base Quality Bias</strong> 
        .. Tests if variant bases tend to occur with a quality bias (T-test). This filter is by default 
        effectively disabled as it is set to 0.
    </div>
<p><strong><i>Note:</i></strong> A fast <a href="htslib.html#annotate">htslib</a> C version of this tool is now available (see <a href="htslib.html#annotate">bcftools annotate</a>).
<div class="usageBox"><span class="usageToggle">(Read more)</span><pre class="usageText">About: Annotates VCF file, adding filters or custom annotations. Requires tabix indexed file with annotations.
   Currently it can annotate ID, QUAL, FILTER and INFO columns, but will be extended on popular demand.
   For examples of user-defined filters see online documentation or examples/filters.txt in vcftools distribution.
Usage: cat in.vcf | vcf-annotate [OPTIONS] &gt; out.vcf
Options:
   -a, --annotations &lt;file.gz&gt;         The tabix indexed file with the annotations: CHR\tFROM[\tTO][\tVALUE]+.
   -c, --columns &lt;list&gt;                The list of columns in the annotation file, e.g. CHROM,FROM,TO,-,QUAL,INFO/STR,INFO/GN. The dash
                                           in this example indicates that the third column should be ignored. If TO is not
                                           present, it is assumed that TO equals to FROM. When REF and ALT columns are present, only
                                           matching lines are annotated.
   -d, --description &lt;file|string&gt;     Header annotation, e.g. key=INFO,ID=HM2,Number=0,Type=Flag,Description='HapMap2 membership'.
                                           The descriptions can be read from a file, one annotation per line.
       --fill-AC-AN                    (Re)Calculate AC and AN tags
       --fill-HWE                      (Re)Calculate HWE, AC and AN tags
       --fill-ICF                      (Re)Calculate Inbreeding Coefficient F, HWE, AC and AN
       --fill-type                     Annotate INFO/TYPE with snp,del,ins,mnp,complex
   -f, --filter &lt;file|list&gt;            Apply filters, list is in the format flt1=value/flt2/flt3=value/etc. If argument to -f is a file,
                                           user-defined filters be applied. See User Defined Filters below.
   -H, --hard-filter                   Remove lines with FILTER anything else than PASS or &quot;.&quot;
   -n, --normalize-alleles             Make REF and ALT alleles more compact if possible (e.g. TA,TAA -&gt; T,TA).
   -r, --remove &lt;list&gt;                 Comma-separated list of tags to be removed (e.g. ID,INFO/DP,FORMAT/DP,FILTER).
   -h, -?, --help                      This help message.
Filters:
	+                           		Apply all filters with default values (can be overriden, see the example below).
	-X                          		Exclude the filter X
	1, StrandBias  FLOAT        		Min P-value for strand bias (INFO/PV4) [0.0001]
	2, BaseQualBias  FLOAT      		Min P-value for baseQ bias (INFO/PV4) [0]
	3, MapQualBias  FLOAT       		Min P-value for mapQ bias (INFO/PV4) [0]
	4, EndDistBias  FLOAT       		Min P-value for end distance bias (INFO/PV4) [0.0001]
	a, MinAB  INT               		Minimum number of alternate bases (INFO/DP4) [2]
	c, SnpCluster  INT1,INT2    		Filters clusters of 'INT1' or more SNPs within a run of 'INT2' bases []
	D, MaxDP  INT               		Maximum read depth (INFO/DP or INFO/DP4) [10000000]
	d, MinDP  INT               		Minimum read depth (INFO/DP or INFO/DP4) [2]
	H, HWE  FLOAT               		Minimum P-value for HWE and F&lt;0 (invokes --fill-HWE) []
	H2, HWE2  FLOAT              		Minimum P-value for HWE (plus F&lt;0) (INFO/AC and INFO/AN or --fill-AC-AN) []
	HG, HWE_G3  FLOAT            		Minimum P-value for HWE and F&lt;0 (INFO/HWE and INFO/G3) []
	q, MinMQ  INT               		Minimum RMS mapping quality for SNPs (INFO/MQ) [10]
	Q, Qual  INT                		Minimum value of the QUAL field [10]
	r, RefN                     		Reference base is N []
	v, VDB  FLOAT               		Minimum Variant Distance Bias (INFO/VDB) [0]
	W, GapWin  INT              		Window size for filtering adjacent gaps [3]
	w, SnpGap  INT              		SNP within INT bp around a gap to be filtered [10]
Examples:
   zcat in.vcf.gz | vcf-annotate -a annotations.gz -d descriptions.txt -c FROM,TO,CHROM,ID,INFO/DP | bgzip -c &gt;out.vcf.gz
   zcat in.vcf.gz | vcf-annotate -f +/-a/c=3,10/q=3/d=5/-D -a annotations.gz -d key=INFO,ID=GN,Number=1,Type=String,Description='Gene Name' | bgzip -c &gt;out.vcf.gz
   zcat in.vcf.gz | vcf-annotate -a dbSNPv132.tab.gz -c CHROM,POS,REF,ALT,ID,-,-,- | bgzip -c &gt;out.vcf.gz
   zcat in.vcf.gz | vcf-annotate -r FILTER/MinDP | bgzip -c &gt;out.vcf.gz
Where descriptions.txt contains:
   key=INFO,ID=GN,Number=1,Type=String,Description='Gene Name'
   key=INFO,ID=STR,Number=1,Type=Integer,Description='Strand'
The file dbSNPv132.tab.gz with dbSNP IDs can be downloaded from
   ftp://ftp.sanger.ac.uk/pub/1000genomes/pd3/dbSNP/
</pre></div>
<div class="usageBox"><span class="usageToggle">(Read even more)</span><pre class="usageText"># Examples of user-defined filters. Edit and run with -f filters.txt.
# The examples below are self-explanatory. Notice the use of the predefined
#  variables ($PASS, $FAIL, $MATCH, $RECORD) and methods (error).


# In this example, a minimum value of AF1=0.1 is required
{
    tag  =&gt; 'INFO/AF1',                       # The VCF tag to apply this filter on
        name =&gt; 'MinAF',                          # The filter ID
        desc =&gt; 'Minimum AF1 [0.01]',             # Description for the VCF header
        test =&gt; sub { return $MATCH &lt; 0.01 ? $FAIL : $PASS },
},


# Filter all indels (presence of INDEL tag is tested)
{
    tag      =&gt; 'INFO/INDEL',
    apply_to =&gt; 'indels',         # Can be one of SNPs, indels, all. Default: [All]
    name     =&gt; 'Indel',
    desc     =&gt; 'INDEL tag present',
    test     =&gt; sub { return $FAIL },
},


# Only loci with enough reads supporting the variant will pass the filter
{
    tag      =&gt; 'INFO/DP4',
    name     =&gt; 'FewAlts',
    desc     =&gt; 'Too few reads supporting the variant',
    apply_to =&gt; 'SNPs',
    test     =&gt; sub {
        if ( !($MATCH =~ /^([^,]+),([^,]+),([^,]+),(.+)$/) )
        {
            error(&quot;Could not parse INFO/DP4: $CHROM:$POS [$MATCH]&quot;);
        }
        if ( 0.1*($1+$2) &gt; $3+$4  ) { return $PASS; }
        return $FAIL;
    },
},

# Example of filtering based on genotype columns and the QUAL column
{
    tag      =&gt; 'FORMAT/PL',
    name     =&gt; 'NoHets',
    desc     =&gt; 'Inbred homozygous mouse, no hets expected',
    apply_to =&gt; 'SNPs',
    test     =&gt; sub {
            for my $pl (@$MATCH)
            {
                my @pls = split(/,/,$pl);
                if ( $pls[1]&lt;$pls[0] &amp;&amp; $pls[1]&lt;$pls[2] ) { return $FAIL; }
            }
        return $PASS;
    },
},


# This example splits the four PV4 values into four tags names PV0, PV1, PV2 and PV3.
#   Note the use of the 'header' key, and the $RECORD and $VCF variables.
{
    header   =&gt; [
                    qq[key=INFO,ID=PV0,Number=1,Type=Float,Description=&quot;P-value for strand bias&quot;],
                    qq[key=INFO,ID=PV1,Number=1,Type=Float,Description=&quot;P-value for baseQ bias&quot;],
                    qq[key=INFO,ID=PV2,Number=1,Type=Float,Description=&quot;P-value for mapQ bias&quot;],
                    qq[key=INFO,ID=PV3,Number=1,Type=Float,Description=&quot;P-value for tail distance bias&quot;]
                ],
    tag      =&gt; 'INFO/PV4',
    name     =&gt; 'SplitPV4',
    desc     =&gt; 'Split PV4',
    apply_to =&gt; 'all',
    test     =&gt; sub {
        my @vals = split(/,/,$MATCH);
        $$RECORD[7] = $VCF-&gt;add_info_field($$RECORD[7],'PV0'=&gt;$vals[0],'PV1'=&gt;$vals[1],'PV2'=&gt;$vals[2],'PV3'=&gt;$vals[3]);
        return $PASS;
    },
},

# Do whatever you want with every record and edit it according to your needs. This silly
#   example removes the tag SILLY in records where ID is set and depth is bigger than 5.
{
    tag      =&gt; 'Dummy',
    test     =&gt; sub {
        if ( $$RECORD[2] eq '.' ) { return $PASS; } # Modify only lines with ID
        my $dp = $vcf-&gt;get_info_field($$RECORD[7],'DP');
        if ( $dp&gt;5 ) { $$RECORD[7] = $VCF-&gt;add_info_field($$RECORD[7],'SILLY'=&gt;undef); }
        return $PASS;
    },
}


# Filter records with the value XY absent or not equal to 42
{
    tag      =&gt; 'Dummy',
    header   =&gt; [
        qq[key=FILTER,ID=XY,Description=&quot;XY not OK&quot;],
    ],
    test     =&gt; sub {
        my $xy      = $VCF-&gt;get_info_field($$RECORD[7],'XY');
        my $is_bad  = ( !defined $xy or $xy!=42 ) ? 1 : 0;
        $$RECORD[6] = $VCF-&gt;add_filter($$RECORD[6],'XY'=&gt;$is_bad);
        return $PASS;
    },
},


# Annotate INFO field with SINGLETON flag when one and only one sample is different from the reference
{
    header   =&gt; [
        qq[key=INFO,ID=SINGLETON,Number=0,Type=Flag,Description=&quot;Only one non-ref sample&quot;],
    ],
    tag      =&gt; 'FORMAT/GT',
    name     =&gt; 'Dummy',
    desc     =&gt; 'Dummy',
    test     =&gt; sub {
        my $nalt = 0;
        for my $gt (@$MATCH)
        {
            my @gt = $VCF-&gt;split_gt($gt);
            for my $allele (@gt)
            {
                if ( $allele ne 0 &amp;&amp; $allele ne '.' ) { $nalt++; last; }
            }
            if ( $nalt&gt;1 ) { last; }
        }
        if ( $nalt==1 ) { $$RECORD[7] = $VCF-&gt;add_info_field($$RECORD[7],'SINGLETON'=&gt;''); }
        return $PASS;
    },
},


# Set genotypes to unknown (&quot;.&quot; or &quot;./.&quot; depending on ploidy) when coverage is low (by Shane McCarthy).
{
    tag      =&gt; 'FORMAT/DP',
    name     =&gt; 'MinSampleDP',
    desc     =&gt; 'Genotypes set to . for samples with DP &lt; 2',
    apply_to =&gt; 'all',
    test     =&gt; sub {
        my $i = 8;
        for my $dp (@$MATCH)
        {
            $i++;
            next unless ($dp&lt;2);
            my @format = split(/:/,$$RECORD[$i]);
            $format[0] = $format[0] =~ /\// ? &quot;./.&quot; : &quot;.&quot;;
            $$RECORD[$i] = join(&quot;:&quot;,@format);
        }
        return $PASS;
    },
},
</pre></div>

<h3><a name="vcf-compare" class="Q">vcf-compare</a></h3>
<p> Compares positions in two or more VCF files and outputs the numbers of
positions contained in one but not the other files; two but not the other files, etc, which
comes handy when generating Venn diagrams. The script also computes numbers
such as nonreference discordance rates (including multiallelic sites), compares
actual sequence (useful when comparing indels), etc.
<br>
</p>
<p class="codebox">vcf-compare -H A.vcf.gz B.vcf.gz C.vcf.gz</p>
<p><strong><i>Note:</i></strong> A fast <a href="htslib.html#stats">htslib</a> C version of this tool is now available (see <a href="htslib.html#stats">bcftools stats</a>).
</p>
<div class="usageBox"><span class="usageToggle">(Read more)</span><pre class="usageText">About: Compare bgzipped and tabix indexed VCF files. (E.g. bgzip file.vcf; tabix -p vcf file.vcf.gz)
Usage: vcf-compare [OPTIONS] file1.vcf file2.vcf ...
       vcf-compare -p plots chr1.cmp chr2.cmp ...
Options:
   -a, --apply-filters                 Ignore lines where FILTER column is anything else than PASS or '.'
   -c, --chromosomes &lt;list|file&gt;       Same as -r, left for backward compatibility. Please do not use as it will be dropped in the future.
   -d, --debug                         Debugging information. Giving the option multiple times increases verbosity
   -g, --cmp-genotypes                 Compare genotypes, not only positions
       --ignore-indels                 Exclude sites containing indels from genotype comparison
   -m, --name-mapping &lt;list|file&gt;      Use with -g when comparing files with differing column names. The argument to this options is a
                                           comma-separated list or one mapping per line in a file. The names are colon separated and must
                                           appear in the same order as the files on the command line.
       --INFO &lt;string&gt; [&lt;int&gt;]         Calculate genotype errors by INFO. Use zero based indecies if field has more than one value. Can be
                                           given multiple times.
   -p, --plot &lt;prefix&gt;                 Create plots. Multiple files (e.g. per-chromosome outputs from vcf-compare) can be given.
   -R, --refseq &lt;file&gt;                 Compare the actual sequence, not just positions. Use with -w to compare indels.
   -r, --regions &lt;list|file&gt;           Process the given regions (comma-separated list or one region per line in a file).
   -s, --samples &lt;list|file&gt;           Process only the listed samples. Excluding unwanted samples may increase performance considerably.
   -t, --title &lt;string&gt;                Title for graphs (see also -p)
   -w, --win &lt;int&gt;                     In repetitive sequences, the same indel can be called at different positions. Consider
                                           records this far apart as matching (be it a SNP or an indel).
   -h, -?, --help                      This help message.
</pre></div>

<h3><a name="vcf-concat" class="Q">vcf-concat</a></h3>
<p>Concatenates VCF files (for example split by chromosome). Note that the input and output VCFs will
have the same number of columns, the script does not merge VCFs by position (see also
<a href="#vcf-merge">vcf-merge</a>).
<p>In the basic mode it does not do anything fancy except for a sanity check that all
files have the same columns.  When run with the -s option, it will perform a partial
merge sort, looking at limited number of open files simultaneously.
<p class="codebox">vcf-concat A.vcf.gz B.vcf.gz C.vcf.gz | gzip -c &gt; out.vcf.gz</p>
<div class="usageBox"><span class="usageToggle">(Read more)</span><pre class="usageText">About: Convenience tool for concatenating VCF files (e.g. VCFs split by chromosome).
   In the basic mode it does not do anything fancy except for a sanity check that all
   files have the same columns.  When run with the -s option, it will perform a partial
   merge sort, looking at limited number of open files simultaneously.
Usage: vcf-concat [OPTIONS] A.vcf.gz B.vcf.gz C.vcf.gz &gt; out.vcf
Options:
   -c, --check-columns              Do not concatenate, only check if the columns agree.
   -f, --files &lt;file&gt;               Read the list of files from a file.
   -p, --pad-missing                Write '.' in place of missing columns. Useful for joining chrY with the rest.
   -s, --merge-sort &lt;int&gt;           Allow small overlaps in N consecutive files.
   -h, -?, --help                   This help message.
</pre></div>

<h3><a name="vcf-consensus" class="Q">vcf-consensus</a></h3>
<p>Apply VCF variants to a fasta file to create consensus sequence.
</p>
<p class="codebox">cat ref.fa | vcf-consensus file.vcf.gz &gt; out.fa</p>
<div class="usageBox"><span class="usageToggle">(Read more)</span><pre class="usageText">Usage: cat ref.fa | vcf-consensus [OPTIONS] in.vcf.gz &gt; out.fa
Options:
   -h, -?, --help                   This help message.
   -H, --haplotype &lt;int&gt;            Apply only variants for the given haplotype (1,2)
   -i, --iupac-codes                Apply variants in the form of IUPAC ambiguity codes
   -s, --sample &lt;name&gt;              If not given, all variants are applied
Examples:
   # Get the consensus for one region. The fasta header lines are then expected
   # in the form &quot;&gt;chr:from-to&quot;.
   samtools faidx ref.fa 8:11870-11890 | vcf-consensus in.vcf.gz &gt; out.fa
</pre></div>

<h3><a name="vcf-convert" class="Q">vcf-convert</a></h3>
<p>Convert between VCF versions, currently from VCFv3.3 to VCFv4.0.
</p>
<p class="codebox">zcat file.vcf.gz | vcf-convert -r reference.fa &gt; out.vcf</p>
<div class="usageBox"><span class="usageToggle">(Read more)</span><pre class="usageText">About: Convert between VCF versions.
Usage: cat in.vcf | vcf-convert [OPTIONS] &gt; out.vcf
Options:
   -r, --refseq &lt;file&gt;              The reference sequence in samtools faindexed fasta file. (Not required with SNPs only.)
   -v, --version &lt;string&gt;           4.0, 4.1, 4.2
   -h, -?, --help                   This help message.
</pre></div>

<h3><a name="vcf-contrast" class="Q">vcf-contrast</a></h3>
<p> A tool for finding differences between groups of samples, useful in trio analysises, cancer genomes etc. 
<p>
In the example below variants with average mapping quality of 30 (<span class="cmd">-f MinMQ=30</span>) 
and minimum depth of 10 (<span class="cmd">-d 10</span>) are considered.
Only novel alleles are reported (<span class="cmd">-n</span>). Then <span class="cmd">vcf-query</span> 
is used to extract the INFO/NOVEL* annotations into a table. Finally the sites are 
sorted by confidence of the site being different in the child (<span class="cmd">-k5,5nr</span>).
</p>
<p class="codebox">
vcf-annotate -f MinMQ=30 file.vcf | vcf-contrast -n +Child -Mother,Father -d 10 -f | vcf-query -f '%CHROM %POS\t%INFO/NOVELTY\t%INFO/NOVELAL\t%INFO/NOVELGT[\t%SAMPLE %GTR %PL]\n' | sort -k3,3nr | head</p>
<div class="usageBox"><span class="usageToggle">(Read more)</span><pre class="usageText">About: Finds differences amongst samples adding NOVELGT, NOVELAL and NOVELTY annotations to INFO field.
       Note that haploid genotypes are internally treated as homozygous diploid genotypes, therefore
       &quot;0/1&quot; and &quot;1&quot; are considered different genotypes.
Usage: vcf-contrast +&lt;list&gt; -&lt;list&gt; [OPTIONS] file.vcf.gz
Options:
   +&lt;list&gt;                             List of samples where unique variant is expected
   -&lt;list&gt;                             List of background samples
   -d, --min-DP &lt;int&gt;                  Minimum depth across all -&lt;list&gt; samples
   -f, --apply-filters                 Skip sites with FILTER column different from PASS or &quot;.&quot;
   -n, --novel-sites                   Print only records with novel genotypes
   -h, -?, --help                      This help message.
Example:
   # Test if any of the samples A,B is different from all C,D,E
   vcf-contrast +A,B -C,D,E -m file.vcf.gz

   # Same as above but printing only sites with novel variants and table output
   vcf-contrast -n +A,B -C,D,E -m file.vcf.gz | vcf-query -f '%CHROM %POS\t%INFO/NOVELTY\t%INFO/NOVELAL\t%INFO/NOVELGT[\t%SAMPLE %GTR %PL]\n'

   # Similar to above but require minimum mapping quality of 20
   vcf-annotate -f MinMQ=20 file.vcf.gz | vcf-contrast +A,B,C -D,E,F -f
</pre></div>

<h3><a name="vcf-filter" class="Q">vcf-filter</a></h3>
<p>Please take a look at <a href="#vcf-annotate">vcf-annotate</a> and <a href="htslib.html#view">bcftools view</a> which does what you are looking for. Apologies for the non-intuitive naming.
<br><strong><i>Note:</i></strong> A fast <a href="htslib.html">HTSlib</a> C version of a filtering tool is now available (see <a href="htslib.html#filter">bcftools filter</a> and
<a href=htslib.html#view>bcftools view</a>).
</p>

<h3><a name="vcf-fix-newlines" class="Q">vcf-fix-newlines</a></h3>
<p> Fixes diploid vs haploid genotypes on sex chromosomes, including the pseudoautosomal regions.
</p>
<div class="usageBox"><span class="usageToggle">(Read more)</span><pre class="usageText">About: Reads in a VCF file with any (commonly used) newline representation and outputs with the
	current system's newline representation.
Usage: vcf-fix-newlines [OPTIONS]
Options:
   -i, --info                      Report if the file is consistent with the current platform based.
   -h, -?, --help                  This help message.
Example:
	vcf-fix-newlines -i file.vcf
	vcf-fix-newlines file.vcf.gz &gt; out.vcf
	cat file.vcf | vcf-fix-newlines &gt; out.vcf
</pre></div>

<h3><a name="vcf-fix-ploidy" class="Q">vcf-fix-ploidy</a></h3>
<p> Fixes diploid vs haploid genotypes on sex chromosomes, including the pseudoautosomal regions.
</p>
<div class="usageBox"><span class="usageToggle">(Read more)</span><pre class="usageText">Usage: cat broken.vcf | vcf-fix-ploidy [OPTIONS] &gt; fixed.vcf
Options:
   -a, --assumed-sex &lt;sex&gt;         M or F, required if the list is not complete in -s
   -l, --fix-likelihoods           Add or remove het likelihoods (not the default behaviour)
   -p, --ploidy &lt;file&gt;             Ploidy definition. The default is shown below.
   -s, --samples &lt;file&gt;            List of sample sexes (sample_name [MF]).
   -h, -?, --help                  This help message.
Default ploidy definition:
   ploidy =&gt;
   {
       X =&gt;
       [
           # The pseudoautosomal regions 60,001-2,699,520 and 154,931,044-155,270,560 with the ploidy 2
           { from=&gt;1, to=&gt;60_000, M=&gt;1 },
           { from=&gt;2_699_521, to=&gt;154_931_043, M=&gt;1 },
       ],
       Y =&gt;
       [
           # No chrY in females and one copy in males
           { from=&gt;1, to=&gt;59_373_566, M=&gt;1, F=&gt;0 },
       ],
       MT =&gt;
       [
           # Haploid MT in males and females
           { from=&gt;1, to =&gt; 16_569, M=&gt;1, F=&gt;1 },
       ],
   }
</pre></div>

<h3><a name="vcf-indel-stats" class="Q">vcf-indel-stats</a></h3>
<p> Calculate in-frame ratio.
</p>
<p><strong><i>Note:</i></strong> A fast <a href="htslib.html#stats">htslib</a> C version of this tool is now available (see <a href="htslib.html#stats">bcftools stats</a>).
<div class="usageBox"><span class="usageToggle">(Read more)</span><pre class="usageText">About: Currently calculates in-frame ratio.
Usage: vcf-indel-stats [OPTIONS] &lt; in.vcf &gt; out.txt
Options:
   -h, -?, --help                   This help message.
   -e, --exons &lt;file&gt;               Tab-separated file with exons (chr,from,to; 1-based, inclusive)
   -v, --verbose
</pre></div>

<h3><a name="vcf-isec" class="Q">vcf-isec</a></h3>
<p> Creates intersections and complements of two or more VCF files. Given multiple
    VCF files, it can output the list of positions which are shared by at least N files,
    at most N files, exactly N files, etc. The first example below outputs positions shared by
    at least two files and the second outputs positions present in the files A but absent
    from files B and C.
</p>
<p class="codebox">
vcf-isec -n +2 A.vcf.gz B.vcf.gz | bgzip -c &gt; out.vcf.gz <br>
vcf-isec -c A.vcf.gz B.vcf.gz C.vcf.gz | bgzip -c &gt; out.vcf.gz
</p>
<p><strong><i>Note:</i></strong> A fast <a href="htslib.html">htslib</a> C version of this tool is now available (see <a href="htslib.html#isec">bcftools isec</a>).
</p>
<div class="usageBox"><span class="usageToggle">(Read more)</span><pre class="usageText">About: Create intersections, unions, complements on bgzipped and tabix indexed VCF or tab-delimited files.
   Note that lines from all files can be intermixed together on the output, which can yield
   unexpected results.
Usage: vcf-isec [OPTIONS] file1.vcf file2.vcf ...
Options:
   -a, --apply-filters                 Ignore lines where FILTER column is anything else than PASS or '.'
   -c, --complement                    Output positions present in the first file but missing from the other files.
   -d, --debug                         Debugging information
   -f, --force                         Continue even if the script complains about differing columns, VCF versions, etc.
   -o, --one-file-only                 Print only entries from the left-most file. Without -o, all unique positions will be printed.
   -n, --nfiles [+-=]&lt;int&gt;             Output positions present in this many (=), this many or more (+), or this many or fewer (-) files.
   -p, --prefix &lt;path&gt;                 If present, multiple files will be created with all possible isec combinations. (Suitable for Venn Diagram analysis.)
   -r, --regions &lt;list|file&gt;           Do only the given regions (comma-separated list or one region per line in a file).
   -t, --tab &lt;chr:pos:file&gt;            Tab-delimited file with indexes of chromosome and position columns. (1-based indexes)
   -w, --win &lt;int&gt;                     In repetitive sequences, the same indel can be called at different positions. Consider
                                           records this far apart as matching (be it a SNP or an indel).
   -h, -?, --help                      This help message.
Examples:
   bgzip file.vcf; tabix -p vcf file.vcf.gz
   bgzip file.tab; tabix -s 1 -b 2 -e 2 file.tab.gz
</pre></div>

<h3><a name="vcf-merge" class="Q">vcf-merge</a></h3>
<p> Merges two or more VCF files into one so that, for example, if two source files
    had one column each, on output will be printed a file with two columns.
    See also <a href="vcf-concat">vcf-concat</a> for concatenating VCFs split by chromosome.
</p>
<p class="codebox">vcf-merge A.vcf.gz B.vcf.gz C.vcf.gz | bgzip -c &gt; out.vcf.gz</p>
<p>Note that this script is not intended for concatenating VCF files. For this, use
<a href="perl_module.html#vcf-concat">vcf-concat</a> instead.
<br><strong><i>Note:</i></strong> A fast <a href="htslib.html">htslib</a> C version of this tool is now available (see <a href="htslib.html#merge">bcftools merge</a>).
</p>
<div class="usageBox"><span class="usageToggle">(Read more)</span><pre class="usageText">About: Merges VCF files by position, creating multi-sample VCFs from fewer-sample VCFs.
   The tool requires bgzipped and tabix indexed VCF files on input. (E.g. bgzip file.vcf; tabix -p vcf file.vcf.gz)
   If you need to concatenate VCFs (e.g. files split by chromosome), look at vcf-concat instead.
Usage: vcf-merge [OPTIONS] file1.vcf file2.vcf.gz ... &gt; out.vcf
Options:
   -c, --collapse &lt;snps|indels|both|any|none&gt;      treat as identical sites with differing alleles [any]
   -d, --remove-duplicates                         If there should be two consecutive rows with the same chr:pos, print only the first one.
   -H, --vcf-header &lt;file&gt;                         Use the provided VCF header
   -h, -?, --help                                  This help message.
   -r, --regions &lt;list|file&gt;                       Do only the given regions (comma-separated list or one region per line in a file).
   -R, --ref-for-missing &lt;string&gt;                  Use the REF allele instead of the default missing genotype. Because it is not obvious
                                                       what ploidy should be used, a user-defined string is used instead (e.g. 0/0).
   -s, --silent                                    Try to be a bit more silent, no warnings about duplicate lines.
   -t, --trim-ALTs                                 If set, redundant ALTs will be removed
</pre></div>

<h3><a name="vcf-phased-join" class="Q">vcf-phased-join</a></h3>
<p> Concatenates multiple overlapping VCFs preserving phasing.
</p>
<div class="usageBox"><span class="usageToggle">(Read more)</span><pre class="usageText">About: The script takes multiple overlapping pre-phased chunks and concatenates them into one VCF
   using heterozygous calls from the overlaps to determine correct phase.
Usage: vcf-phased-join [OPTIONS] A.vcf B.vcf C.vcf
Options:
   -j, --min-join-quality &lt;num&gt;    Quality threshold for gluing the pre-phased blocks together [10]
   -l, --list &lt;file&gt;               List of VCFs to join.
   -o, --output &lt;file&gt;             Output file name. When &quot;-&quot; is supplied, STDOUT and STDERR will be used
   -q, --min-PQ &lt;num&gt;              Break pre-phased segments if PQ value is lower in input VCFs [0.6]
   -h, -?, --help                  This help message
</pre></div>

<h3><a name="vcf-query" class="Q">vcf-query</a></h3>
<p> Powerful tool for converting VCF files into format defined by the user.
Supports retrieval of subsets of positions, columns and fields.
</p>
<p class="codebox">
vcf-query file.vcf.gz 1:10327-10330<br>
vcf-query file.vcf -f '%CHROM:%POS %REF %ALT [ %DP]\n' 
</p>
<p>
<strong><i>Note:</i></strong> A fast <a href="htslib.html">htslib</a> C version of this tool is now available (see <a href="htslib.html#query">bcftools query</a>).
</p>
<div class="usageBox"><span class="usageToggle">(Read more)</span><pre class="usageText">Usage: vcf-query [OPTIONS] file.vcf.gz
Options:
   -c, --columns &lt;file|list&gt;           List of comma-separated column names or one column name per line in a file.
   -f, --format &lt;string&gt;               The default is '%CHROM:%POS\t%REF[\t%SAMPLE=%GT]\n'
   -l, --list-columns                  List columns.
   -r, --region chr:from-to            Retrieve the region. (Runs tabix.)
       --use-old-method                Use old version of API, which is slow but more robust.
   -h, -?, --help                      This help message.
Expressions:
   %CHROM          The CHROM column (similarly also other columns)
   %GT             Translated genotype (e.g. C/A)
   %GTR            Raw genotype (e.g. 0/1)
   %INFO/TAG       Any tag in the INFO column
   %LINE           Prints the whole line
   %SAMPLE         Sample name
   []              The brackets loop over all samples
   %*&lt;A&gt;&lt;B&gt;        All format fields printed as KEY&lt;A&gt;VALUE&lt;B&gt;
Examples:
   vcf-query file.vcf.gz 1:1000-2000 -c NA001,NA002,NA003
   vcf-query file.vcf.gz -r 1:1000-2000 -f '%CHROM:%POS\t%REF\t%ALT[\t%SAMPLE:%*=,]\n'
   vcf-query file.vcf.gz -f '[%GT\t]%LINE\n'
   vcf-query file.vcf.gz -f '[%GT\ ]%LINE\n'
   vcf-query file.vcf.gz -f '%CHROM\_%POS\t%INFO/DP\t%FILTER\n'
Notes:
   Please use `bcftools query` instead, this script will not be supported in future.
</pre></div>

<h3><a name="shuffle" class="Q">vcf-shuffle-cols</a></h3>
<p>Reorder columns</p>
<p class="codebox">vcf-shuffle-cols -t template.vcf.gz file.vcf.gz > out.vcf
</p>
<div class="usageBox"><span class="usageToggle">(Read more)</span><pre class="usageText">About: Reorder columns to match the order in the template VCF.
Usage: vcf-shuffle-cols [OPTIONS] -t template.vcf.gz file.vcf.gz &gt; out.vcf
Options:
   -t, --template &lt;file&gt;            The file with the correct order of the columns.
   -h, -?, --help                   This help message.
</pre></div>

<h3><a name="vcf-sort" class="Q">vcf-sort</a></h3>
<p>Sort a VCF file.
</p>
<p class="codebox">vcf-sort file.vcf.gz</p>
<div class="usageBox"><span class="usageToggle">(Read more)</span><pre class="usageText">Usage: vcf-sort &gt; out.vcf
       cat file.vcf | vcf-sort &gt; out.vcf
Options:
   -c, --chromosomal-order       Use natural ordering (1,2,10,MT,X) rather then the default (1,10,2,MT,X). This requires
                                     new version of the unix &quot;sort&quot; command which supports the --version-sort option.
   -p, --parallel &lt;int&gt;          Change the number of sorts run concurrently to &lt;int&gt;
   -t, --temporary-directory     Use a directory other than /tmp as the temporary directory for sorting.
   -h, -?, --help                This help message.
</pre></div>

<h3><a name="vcf-stats" class="Q">vcf-stats</a></h3>
<p>Outputs some basic statistics: the number of SNPs, indels, etc.
</p>
<p class="codebox">vcf-stats file.vcf.gz</p>
<p><strong><i>Note:</i></strong> A fast <a href="htslib.html#stats">htslib</a> C version of this tool is now available (see <a href="htslib.html#stats">bcftools stats</a>).
<div class="usageBox"><span class="usageToggle">(Read more)</span><pre class="usageText">Use of qw(...) as parentheses is deprecated at /usr/share/perl5/VcfStats.pm line 428.
Use of qw(...) as parentheses is deprecated at /usr/share/perl5/VcfStats.pm line 433.
Usage: vcf-stats [OPTIONS] file.vcf.gz
Options:
   -d, --dump &lt;file&gt;                           Take an existing dump file and recreate the files (works with -p)
   -f, --filters &lt;filter1,filter2&gt;             List of filters such as column/field (any value), column/field=bin:max (cluster in bins),column/field=value (exact value)
   -p, --prefix &lt;dir/string&gt;                   Prefix of output files. If slashes are present, directories will be created.
   -s, --samples &lt;list&gt;                        Process only the listed samples, - for none. Excluding unwanted samples may increase performance considerably.
   -h, -?, --help                              This help message.

Examples:
   # Calculate stats separately for the filter field, quality and non-indels
   vcf-stats file.vcf.gz -f FILTER,QUAL=10:200,INFO/INDEL=False -p out/

   # Calculate stats for all samples
   vcf-stats file.vcf.gz -f FORMAT/DP=10:200 -p out/

   # Calculate stats only for the sample NA00001
   vcf-stats file.vcf.gz -f SAMPLE/NA00001/DP=1:200 -p out/

   vcf-stats file.vcf.gz &gt; perl.dump
</pre></div>

<h3><a name="vcf-subset" class="Q">vcf-subset</a></h3>
<p>Remove some columns from the VCF file.
</p>
<p class="codebox">vcf-subset -c NA0001,NA0002 file.vcf.gz | bgzip -c &gt; out.vcf.gz</p>
<p>
<strong><i>Note:</i></strong> A fast <a href="htslib.html">HTSlib</a> C version of this tool is now available (see <a href="htslib.html#view">bcftools view</a>).
</p>
<div class="usageBox"><span class="usageToggle">(Read more)</span><pre class="usageText">Usage: vcf-subset [OPTIONS] in.vcf.gz &gt; out.vcf
Options:
   -a, --trim-alt-alleles           Remove alternate alleles if not found in the subset
   -c, --columns &lt;string&gt;           File or comma-separated list of columns to keep in the vcf file. If file, one column per row
   -e, --exclude-ref                Exclude rows not containing variants.
   -f, --force                      Proceed anyway even if VCF does not contain some of the samples.
   -p, --private                    Print only rows where only the subset columns carry an alternate allele.
   -r, --replace-with-ref           Replace the excluded types with reference allele instead of dot.
   -t, --type &lt;list&gt;                Comma-separated list of variant types to include: ref,SNPs,indels,MNPs,other.
   -u, --keep-uncalled              Do not exclude rows without calls.
   -h, -?, --help                   This help message.
Examples:
   cat in.vcf | vcf-subset -r -t indels -e -c SAMPLE1 &gt; out.vcf
</pre></div>

<h3><a name="vcf-tstv" class="Q">vcf-tstv</a></h3>
<p>A lightweight script for quick calculation of <a href="http://en.wikipedia.org/wiki/Transition_%28genetics%29">Ts</A>/<A href="http://en.wikipedia.org/wiki/Transversion">Tv</A> ratio.
</p>
<p class="codebox">cat file.vcf | vcf-tstv</p>
<p><strong><i>Note:</i></strong> A fast <a href="htslib.html#stats">htslib</a> C version of this tool is now available (see <a href="htslib.html#stats">bcftools stats</a>).
<div class="usageBox"><span class="usageToggle">(Read more)</span><pre class="usageText">Usage: cat file.vcf | vcf-tstv
Options:
   -h, -?, --help                  This help message.
</pre></div>

<h3><a name="vcf-to-tab" class="Q">vcf-to-tab</a></h3>
<p>A simple script which converts the VCF file into a tab-delimited text file listing the
    actual variants instead of ALT indexes.
</p>
<p class="codebox">zcat file.vcf.gz | vcf-to-tab &gt; out.tab</p>
<div class="usageBox"><span class="usageToggle">(Read more)</span><pre class="usageText">Usage: vcf-to-tab [OPTIONS] &lt; in.vcf &gt; out.tab
Options:
   -h, -?, --help                   This help message.
   -i, --iupac                      Use one-letter IUPAC codes
Notes:
   Please use `bcftools query` instead, this script will not be supported in future.
</pre></div>

<h3><a name="vcf-validator" class="Q">vcf-validator</a></h3>
<p>
</p>
<p class="codebox">vcf-validator file.vcf.gz</p>
<div class="usageBox"><span class="usageToggle">(Read more)</span><pre class="usageText">Usage: vcf-validator [OPTIONS] file.vcf.gz
Options:
   -d, --duplicates                 Warn about duplicate positions.
   -u, --unique-messages            Output all messages only once.
   -h, -?, --help                   This help message.
</pre></div>

<h3><a name="Vcf.pm" class="Q">Vcf.pm</a></h3>
<p>For examples how to use the Perl API, it is best to look at some of the simpler scripts, 
for example <span class="cmd">vcf-to-tab</span>. The detailed documentation can be obtained by running</p>
<p class="codebox">perldoc Vcf.pm</p>
</section>
      <footer>
        <p><small>Hosted on <a href="https://pages.github.com">GitHub Pages</a></small></p>
        <p>Copyright 2015 &copy; VCFtools</p>
      </footer>
    </div>
    <!--[if !IE]><script>fixScale(document);</script><![endif]-->
		          <script type="text/javascript">
            var gaJsHost = (("https:" == document.location.protocol) ? "https://ssl." : "http://www.");
            document.write(unescape("%3Cscript src='" + gaJsHost + "google-analytics.com/ga.js' type='text/javascript'%3E%3C/script%3E"));
          </script>
          <script type="text/javascript">
            try {
              var pageTracker = _gat._getTracker("UA-272183-4");
            pageTracker._trackPageview();
            } catch(err) {}
          </script>

  </body>
</html>

